compiler/src/dotty/tools/dotc/transform/ArrayApply.scala

```diff
@@ -0,0 +1,53 @@
+package dotty.tools.dotc
+package transform
+
+import core._
+import MegaPhase._
+import Contexts.Context
+import Symbols._
+import Types._
+import StdNames._
+import ast.Trees._
+import dotty.tools.dotc.ast.tpd
+
+
+/** This phase rewrites calls to `Array.apply` to primitive array instantion.
+ *
+ *  Transforms `scala.Array.apply([....])` and `scala.Array.apply(..., [....])` into `[...]`
+ */
+class ArrayApply extends MiniPhase {
+  import tpd._
+
+  override def phaseName: String = "arrayApply"
+
+  override def transformApply(tree: tpd.Apply)(implicit ctx: Context): tpd.Tree = {
+    if (tree.symbol.name == nme.apply && tree.symbol.owner == defn.ArrayModule) { // Is `Array.apply`
+      tree.args match {
+        case CleanTree(Apply(wrapRefArrayMeth, (seqLit: tpd.JavaSeqLiteral) :: Nil)) :: ct :: Nil
+            if defn.WrapArrayMethods().contains(wrapRefArrayMeth.symbol) && elideClassTag(ct) =>
+          seqLit
+
+        case elem0 :: CleanTree(Apply(wrapRefArrayMeth, (seqLit: tpd.JavaSeqLiteral) :: Nil)) :: Nil
+            if defn.WrapArrayMethods().contains(wrapRefArrayMeth.symbol) =>
+          tpd.JavaSeqLiteral(elem0 :: seqLit.elems, seqLit.elemtpt)
+
+        case _ =>
+          tree
+      }
+
+    } else tree
+  }
+
+  // Only optimize when classtag is `ClassTag.apply` or `ClassTag.{Byte, Boolean, ...}`
```

Comment:
ClassTag.apply is a method that takes a parameter, so I don't see how it's going to end up being checked here, also someone could write `ClassTag.apply({???; classOf[Foo])`, so you can't optimize it away in general.

Reply:
Good point. I will have to inspect the tree to make sure it can be optimized away.

Reply:
A safe subset is to only optimize if the argument to `ClassTag.apply` is a tree of the form `Literal(Constant(_: Type))` (this requires your phase to be after https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/ClassOf.scala)